Integrated functionality from volcano into vesuvius

SuperOptimizer opened the pull request:
This PR integrates the current functionality from https://github.com/SuperOptimizer/volcano in vesuvius, along with a few tests. 

I wasn't quite sure of the best way to integrate everything from volcano, so I just manually amalgamated all of the current functionality into vesuvius.h, since volcano is composed of a bunch of small .h files. I had to comment out a few things to try and meet the deadline, including 

- zarr support
- compression support in nrrd

I'll get to these quickly; I'm just running into odd errors with cblosc2 compression and decompression that I've not been able to track down yet.

SuperOptimizer committed:
introduce new functionality from volcano

SuperOptimizer committed:
Create example2.c

SuperOptimizer committed:
Create runtests.c

ubc-nvining commented on runtests.c:
```diff
@@ -0,0 +1,236 @@
+int testzarr() {
+  printf("zarr support not yet implemented\n");
+  return 0;
+  printf("%s\n",__FUNCTION__);
```
I would favour commenting this block out, rather than having the return 0; do all the work.

ubc-nvining commented on vesuvius-c.h:
```diff
@@ -820,4 +820,3366 @@ void reset_mesh_origin_to_roi(TriangleMesh *mesh, const RegionOfInterest *roi) {
+#endif
+
+typedef enum errcode {
+  SUCCESS = 0,
```
you might put a wart on these - VESUVIUS_SUCCESS, VESUVIUS_FAILURE - just in case somebody else tries to use the reserved words.

ubc-nvining commented on vesuvius-c.h:
```diff
@@ -820,4 +820,3366 @@ void reset_mesh_origin_to_roi(TriangleMesh *mesh, const RegionOfInterest *roi) {
+#endif
+
+
+#ifdef VESUVIUS_HEADER_ONLY
```
I don't object to this, but I am curious as to why it was done this way. What's the benefit of using the __attribute__ over static which gets you... a fair amount of the way, and is sort of the traditional thing to do for a single-header lib?

Usually the way these things are done is everybody includes the header, one file in the project #defines something like VESUVIUS_IMPL 1 and off we go to the races. See Sean Barrett's libraries and commentary.

SuperOptimizer replied:
The PUBLIC / PRIVATE #defines are trying to support a couple different ways of using the code:

1) as a static inline header only library where the implementation is compiled into each including .c file
2) as a VESUVIUS_IMPL style define where the implementation is only in one .c file and the header is just #ifdefed as a bunch of prototypes.
3) as a dynamic library, where the public methods all have to be exported by the .c file and the header contains no inline public methods

I didn't quite finish out 2, since I'd still need to do all of the #ifdefing to include/disclude function implementations, but marking everything as `static inline` solves the problem of the user needing to `#define VESUVIUS_IMPLEMENTATION` in a single .c file.

Then we can use just PUBLIC/PRIVATE to hide the linkage and toggle `static inline` annotation as needed.

ubc-nvining commented on vesuvius-c.h:
```diff
@@ -820,4 +820,3366 @@ void reset_mesh_origin_to_roi(TriangleMesh *mesh, const RegionOfInterest *roi) {
+}
+
+
+static const s32 edgeTable[256]={
```
So... is this static or is this PRIVATE?

ubc-nvining commented on vesuvius-c.h:
```diff
@@ -820,4 +820,3366 @@ void reset_mesh_origin_to_roi(TriangleMesh *mesh, const RegionOfInterest *roi) {
+0xf00, 0xe09, 0xd03, 0xc0a, 0xb06, 0xa0f, 0x905, 0x80c,
+0x70c, 0x605, 0x50f, 0x406, 0x30a, 0x203, 0x109, 0x0   };
+
+static const s32 triTable[256][16] =
```
etc

ubc-nvining reviewed (commented):
In general, excellent work; well done. This is exactly what I wanted. I have some small concerns regarding style:

- I would put the Curl dependency behind a #define so that it can be used without it (if you have issues linking versus Curl)
- I'm not wild about the PRIVATE stuff, which also seems to be used inconsistently in a couple of places, most notably the marching cubes tables. I feel like I would prefer to use the standard model that STB/IMGUI uses where you have every file include the header; and then one file in the main project defines some macro to actually ensure that when the header is included, all the code is dumped into it.
- I might put a wart on your function names, just because there might be conflicts. Even just v_ will do the job.

Otherwise, LGTM.

SuperOptimizer committed:
add 3d tiff file for testing

SuperOptimizer committed:
various updates for vesuvius-c.h

- remove public / private in favor of a #define VESUVIUS_IMPL switch
- remove static inline from functions and only support prototypes in headers and no inline code
- slight refactoring to allow for the previous two points
- put curl and zarr support behind a compile switch
- remove error code enum and just use 0 / 1 / -1
- move static marchign cubes data into the only utilizing function

SuperOptimizer committed:
add a simple cmakelists and a few examples

SuperOptimizer commented:
I've made some edits

- I nixxed all of the public / private / static inline stuff in favor of a VESUVIUS_IMPL define. Now, if you do define it, then you get the implementation in one .c file, otherwise you just get the function prototypes. Since the implementation is now confined to a single .c, we don't need to mark everything as static inline
- I moved zarr support and curl support behind VESUVIUS_ZARR_IMPL and VESUVIUS_CURL_IMPL as these both require external dependencies
- I moved the mesher static data into the one function that actually needed it
- I nixxed the err code stuff and forced everything to return 0 / a pointer for success and non zero / NULL for failure
- I made a couple edits for tests and examples, but I'm still working on getting these in a more complete state.

I've not moved everything into one namespace yet. I'm considering the best way to do that. I don't really want to move _every single function_ into a `vs_` for vesuvius namespace. I'd prefer to keep components under their own namespace, f.ex. put all tiff functionality under `tiff_` or all nrrd under `nrrd_` and then any vesuvius specific code, like wiring all the libs together, would then go under `vs_`. 

I would like to have some way to differentiate between the "public" API and any private / helper functions, though, so users aren't accidentally using functionality that isn't supposed to be exported. I'll either just prefix them with a `_` or maybe use two underscores after the namespace for private, and one for public, like `tiff_read` but `tiff__super_secret_read_helper`

SuperOptimizer committed:
wart all of the functions. vs_ is used for public functions while vs__ is used for private functions

SuperOptimizer commented:
I've warted all of the _functions_. If I need to wart all of the structs then I can do that as well.

I went with `vs_` for the public functions and `vs__` for the private functions; I believe that `stb` does something similar. For example I've got some utilities for a few string or OS methods that I don't want end users using so I marked those as `vs__` as well as helper functions for module specific things, etc.

SuperOptimizer committed:
A couple fixes in volume loading and fix tests for the new namespaces

ubc-nvining commented:
"I would like to have some way to differentiate between the "public" API and any private / helper functions, though, so users aren't accidentally using functionality that isn't supposed to be exported."

I mean... why not just not export it? The static keyword works in C in the VESUVIUS_IMPL section just fine.

SuperOptimizer commented:
> I mean... why not just not export it

I usually like to keep the public and private APIs separate, but with this being a single header library it's a bit harder to organize the code that way.

But now that prototypes and implementation are toggled by VESUVIUS_IMPL, I can keep the private APIs and prototypes in the VESUVIUS_IMPL section only so that the user doesn't get the prototypes while #including "vesuvius-c.h" without the VESUVIUS_IMPL define. 

I'll make that change

SuperOptimizer committed:
private APIs should only have prototypes within the VESUVIUS_IMPL define block

SuperOptimizer committed:
make private APIs static

SuperOptimizer commented:
Ok, I moved the private API prototypes into the #VESUVIUS_IMPL define block and marked the prototypes and implementations for them as static.

ubc-nvining commented:
Yeah, happy to merge that, I think. Cheers!
